### ISSUE TITLE
bug: test.py (BaseTransform)

### ISSUE THREAD

**Issue by AceCoooool** [amdegroot/ssd.pytorch#31]
line41: `x = torch.from_numpy(transform(img)[0]).permute(2, 0, 1)`  is not change the bgr to rgb. It's not equal to the `dataset = VOCDetection(args.voc_root, [('2007', set_type)], BaseTransform(300, dataset_mean), AnnotationTransform())` （it change bgr to rgb）.

So, I think it's better to add change the line138 `img = img[:, :, (2, 1, 0)]` in voc0712 to the base_transform function's  * (The results will not change too much if we set vis_threshold=0.6, however in the eval.py, if we use BaseTransform out the dataset, it will change the mAP)

**Comment by AceCoooool** [comment]
https://s3.amazonaws.com/amdegroot-models/ssd_300_VOC0712.pth
channels: RGB   --->mAP=0.7247
channels: BGR   --->mAP=0.7718

**Comment by alexkoltun** [comment]
You are correct, the thing is that the original caffe implementation expects the input to be in BGR, the model you use is converted from the caffe model.
On the other hand, the natively trained model, was trained with RGB inputs (I know because I trained it), so it expects the input to be RGB, and it created the confusion, I hope that in a near feature I will be able to train a new model that will get the inputs as BGR, once ready I will send it to Max so he will be able to publish it.

**Comment by AceCoooool** [comment]
Thank you for your reply, I will close this issue.

**Comment by JappaB** [comment]
Hi guys, 
I'm putting this here as you guys seem to have taken a look at the permutation of the color channels as well..

Do you guys @AceCoooool or @alexkoltun  (or maybe @amdegroot) understand why the image is permuted (2,0,1) in `pull_item` in voc0712.py?

cv2 reads the image as BGR, the base VGG16 model is trained in RGB. But in `pull_item` the images is first transformed to RGB, then later the permute (2,0,1) is used, making it BRG... 

Code at pull item:
`    def pull_item(self, index):
        img_id = self.ids[index]

        target = ET.parse(self._annopath % img_id).getroot()
        img = cv2.imread(self._imgpath % img_id)
        height, width, channels = img.shape

        if self.target_transform is not None:
            target = self.target_transform(target, width, height)

        if self.transform is not None:
            target = np.array(target)
            img, boxes, labels = self.transform(img, target[:, :4], target[:, 4])
            # to rgb
            img = img[:, :, (2, 1, 0)]
            # img = img.transpose(2, 0, 1)
            target = np.hstack((boxes, np.expand_dims(labels, axis=1)))
        return torch.from_numpy(img).permute(2, 0, 1), target, height, width
 `

A Similarly something strange happens in test.py, where the image is just loaded (in BGR) and then transformed with permute(2,0,1), now thus to RBG..

Both channel orders  don't make sense. RGB would make sense and BGR would make sense, depending on which basenet you use (e.g. VGG16 weights from pytorch modelzoo (RGB) or VGG16 weights from the link in this repo: BGR)

**Comment by eleflea** [comment]
`img.transpose(2, 0, 1)` transpose image from (h, w, c) to (c, h, w).

**Comment by JappaB** [comment]
Woooow, I feel so stupid right now. Hahaha thanks a lot @eleflea.
